server/src/instant/db/instaql.clj

```diff
@@ -1967,53 +1977,163 @@
-                              where-conds (rule-wheres->where-conds (:attrs ctx) etype where-clauses)]
-                          (tracer/add-data! {:attributes {:wheres where-clauses
-                                                          :short-circuit? short-circuit?}})
-                          (assoc acc etype {:short-circuit? short-circuit?
-                                            :wheres where-clauses
-                                            :where-conds where-conds}))
-                        (catch Exception e
-                          (tracer/add-data! {:attributes {:exception (.getMessage e)
-                                                          :ex-data (ex-data e)}})
-                          (assoc acc etype {:recheck? true}))))
-                    acc))
-                {}
-                referenced-etypes)))))
+      (let [{:keys [referenced-etypes]} (instaql-query->patterns ctx o)
+            programs (keep (fn [etype]
+                             (rule-model/get-program! rules etype "view"))
+                           referenced-etypes)
+
+            program-results
+            (tracer/with-span! {:name "instaql/get-all-where-clauses"}
+              (cel/get-all-where-clauses ctx rule-params programs))
+            rule-wheres
+            (reduce-kv
+             (fn [acc etype result]
+               (try
+                 (when-let [t (:thrown result)]
+                   (throw t))
+                 (when (seq (:where-clauses result))
+                   ;; Ensures we got valid where clauses back
+                   (rule-wheres->where-conds (:attrs ctx)
+                                             etype
+                                             (:where-clauses result)))
+                 (assoc acc etype {:short-circuit? (:short-circuit? result)
+                                   :where-clauses (:where-clauses result)})
+                 (catch Exception e
+                   (tracer/with-span!
+                       {:name "instaql/rule-where-exception"
+                        :attributes {:code (:code (rule-model/get-program! rules
+                                                                           etype
+                                                                           "view"))
+                                     :error e}}
+                     acc))))
+             {}
+             program-results)]
+        (tracer/add-data! {:attributes {:rule-wheres rule-wheres}})
+        rule-wheres))))
+
+(defn update-where-with-rule-refs [initial-etype attrs rule-wheres where k v]
+  (loop [[segment & rest-path] (string/split (name k) #"\.")
+         current-path [segment]
+         etype initial-etype
+         where (assoc where k v)]
+    (if-let [segment-etype (link-etype attrs etype segment)]
+      (let [rule-where (if (-> rule-wheres (get segment-etype) :short-circuit?)
+                         {:id (str (random-uuid))}
```

Comment:
This is kind of a hack--we know that the rule should always return false, so we add a where clause that we know won't match (id == random-uuid). We might want to add some kind of special `never` operator instead, which could save us some querying.